ckanext/switzerland/logic.py

```diff
@@ -204,3 +221,111 @@ def highlight(term, q):
     except pysolr.SolrError as e:
         log.exception('Could not load suggestions from solr: %s' % e)
     raise ActionError('Error retrieving suggestions from solr')
+
+
+def ogdch_xml_upload(context, data_dict):
+    data = data_dict.get('data')
+    org_id = data_dict.get('organization')
+
+    upload = uploader.get_uploader('dataset_xml')
+    upload.update_data_dict(data, 'dataset_xml',
+                            'file_upload', 'clear_upload')
+    upload.upload()
+
+    dataset_filename = data.get('dataset_xml')
+
+    if not dataset_filename:
+        h.flash_error('Error uploading file.')
+        return
+
+    data_rdfgraph = rdflib.ConjunctiveGraph()
+    profile = SwissDCATAPProfile(data_rdfgraph)
+
+    try:
+        data_rdfgraph.parse(os.path.join(
+            upload.storage_path,
+            dataset_filename
+        ), "xml")
+    except (RDFParserException, SAXParseException) as e:
+        h.flash_error(
+            'Error parsing the RDF file during dataset import: {0}'
+            .format(e))
+
+    for dataset_ref in data_rdfgraph.subjects(RDF.type, DCAT.Dataset):
+        dataset_dict = {}
+        profile.parse_dataset(dataset_dict, dataset_ref)
+        dataset_dict['owner_org'] = org_id
+
+        _create_or_update_dataset(dataset_dict)
+
+
+def _create_or_update_dataset(dataset):
+    context = {}
+    user = tk.get_action('get_site_user')({'ignore_auth': True}, {})
+    context.update({'user': user['name']})
+
+    tk.check_access('package_show', {})
+
+    harvester = SwissDCATRDFHarvester()
+    name = harvester._gen_new_name(dataset['title'])
+
+    package_plugin = lib_plugins.lookup_package_plugin('dataset')
+
+    try:
```

Comment:
I just feel this is a very long try statement. Would it be possible to refactor this using subfunctions? These very long try statements are hard to debug, when things go wrong, because you never know where it stopped. Also with subfunctions the comments lines might become unnecessary. WDYT?